provider.go

```diff
@@ -832,7 +839,7 @@ func (p *provider) Delete(ctx context.Context, req *rpc.DeleteRequest) (*emptypb
 }
 
 func (p *provider) Construct(pctx context.Context, req *rpc.ConstructRequest) (*rpc.ConstructResponse, error) {
-	return comProvider.Construct(pctx, req, p.host.EngineConn(),
+	return comProvider.Construct(putEmbeddedMap(pctx), req, p.host.EngineConn(),
```

Comment:
If we moved `CtxFromPulumiContext` into this file, I think we could remove `embeddedData` and associated publicly exported `GetEmbeddedData` and `PutEmbeddedData`.

Then, the implementation here would look something like the following.

There'd be a `type pkgContextKey struct{}` key.

And inside `Construct` here would look something like:

```go
func (p *provider) Construct(pctx context.Context, req *rpc.ConstructRequest) (*rpc.ConstructResponse, error) {
	packageContext := &pkgContext{}
	pctx = context.WithValue(pctx, pkgContextKey{}, packageContext)
	packageContext.Context = pctx

	return comProvider.Construct(pctx, req, p.host.EngineConn(),
	...
}
```

And the implementation of `CtxFromPulumiContext` would look something like:

```go
func CtxFromPulumiContext(ctx *pulumi.Context) p.Context {
	v := ctx.Context().Value(pkgContextKey{})
 	contract.Assertf(v != nil,
 		"CtxFromPulumiContext must be called on the pulumi.Context passed in to Construct")
 	return v.(p.Context)
}
```

Reply:
I removed `GetEmbeddedData(ctx, key)` function. You can now just call `ctx.Value(key)`. I'm concerned with pre embedding the `provider.Context`, because middleware layers might change it.